check-submodules.sh

```diff
@@ -0,0 +1,146 @@
+#!/bin/bash
+set -euo pipefail
+IFS=$'\n\t'
+
+SCRIPT_ROOT="$(cd -P "$( dirname "$0" )" && pwd)"
+SUBMODULES=$SCRIPT_ROOT/.gitmodules
+
+# has the user answered "yes to all" to init'ing submodules?
+answered_all=0
+
+# ask for confirmation and initialize a submodule if approved.
+init_submodule () {
+  module=$1
+  done=0
+  while [ $done == 0 ]; do
+    echo "warning: submodules $module does not appear to be initialized."
+    read -p "Should I initialize it for you? [Yes/no/all/quit] " -n 1 -r
+    echo
+    if [[ $REPLY =~ ^[Aa]$ ]]; then
+      answered_all=1
+      done=1
+      git submodule update --init --recursive $module
+    elif [[ $REPLY =~ ^[Qq]$ ]]; then
+      exit 0
+    elif [[ $REPLY =~ ^[Nn]$ ]]; then
+      done=1
+    elif [[ $REPLY == "" || $REPLY == " " || $REPLY =~ ^[Yy]$ ]]; then
+      done=1
+      git submodule update --init --recursive $module
+    else
+      echo "didn't understand that ($REPLY)"
+    fi
+  done
+}
+
+# update a submodule to an expected commit.  We give different messages
+# for being ahead vs being behind or diverged from the expected commit,
+# so this function is parameterized.
+fix_submodule () {
+  path=$1
+  expected=$2
+  actual=$3
+  msg=$4
+  prompt=$5
+  done=0
+  while [ $done == 0 ]; do
+    echo -e "$msg"
+    read -p "$prompt " -n 1 -r
+    echo
+    if [[ $REPLY =~ ^[Qq]$ ]]; then
+      exit 1
+    elif [[ $REPLY =~ ^[Yy]$ ]]; then
+	  # check if we have this commit locally and can skip the fetch
+      git cat-file -e $expected^{commit} 2>/dev/null || exitCode=$?
+      if [ $exitCode != 0 ]; then
+        git fetch
+      fi
+      exitCode=0
+	  # double-check, we should have the commit now unless something
+	  # weird is going on.
+      git cat-file -e $expected^{commit} 2>/dev/null || exitCode=$?
+      if [ $exitCode != 0 ]; then
+        echo "error: commit $expected was not found in $path"
+        echo "Are you using a custom remote for this submodule?  You may need to pick up changes from upstream."
+        echo "Canceling remainder of checks."
+        exit 1
+      fi
+      git checkout $expected
+      done=1
+    elif [[ $REPLY == "" || $REPLY == " " || $REPLY =~ ^[Nn]$ ]]; then
+      done=1
+    else
+      echo "didn't understand that ($REPLY)"
+    fi
+  done
+}
+
+# clean a submodule if the user approves.
+clean_submodule() {
```

Comment:
I think it would make sense to have an "All" option on clean as well.

Reply:
Since the `submodule foreach`s are independent this is a little ugly, but I agree that this would be useful.  Let me know if this seems too hacky or if you have a better solution.